File: src/preferences.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import type {
  RemoteTerminalAgentPreferencesData,
  RemoteTerminalQuickCommandData,
  RemoteTerminalShortcutData,
  RemoteTerminalShortcutModifier,
} from "@rtc/protocol";

interface PersistedPreferencesFile {
  version: number;
  defaultWorkingDirectory?: string;
  shortcuts?: RemoteTerminalShortcutData[];
  quickCommands?: RemoteTerminalQuickCommandData[];
}

const PREFERENCES_FILE_VERSION = 1;

function cloneShortcut(shortcut: RemoteTerminalShortcutData): RemoteTerminalShortcutData {
  return {
    ...shortcut,
    sequence: [...shortcut.sequence],
    modifiers: shortcut.modifiers ? [...shortcut.modifiers] : undefined,
  };
}

function cloneQuickCommand(command: RemoteTerminalQuickCommandData): RemoteTerminalQuickCommandData {
  return {
    ...command,
  };
}

function clonePreferences(preferences: RemoteTerminalAgentPreferencesData): RemoteTerminalAgentPreferencesData {
  return {
    defaultWorkingDirectory: preferences.defaultWorkingDirectory,
    shortcuts: preferences.shortcuts.map(cloneShortcut),
    quickCommands: preferences.quickCommands.map(cloneQuickCommand),
  };
}

function sanitizeModifiers(value: unknown): RemoteTerminalShortcutModifier[] | undefined {
  if (!Array.isArray(value)) return undefined;

  const modifiers = value.filter(
    (item): item is RemoteTerminalShortcutModifier =>
      item === "ctrl" || item === "alt" || item === "shift" || item === "meta",
  );

  return modifiers.length > 0 ? modifiers : undefined;
}

function sanitizeShortcuts(value: unknown): RemoteTerminalShortcutData[] {
  if (!Array.isArray(value)) return [];

  const shortcuts: RemoteTerminalShortcutData[] = [];

  for (const item of value) {
    if (!item || typeof item !== "object") continue;

    const candidate = item as Partial<RemoteTerminalShortcutData>;
    const id = typeof candidate.id === "string" ? candidate.id.trim() : "";
    const label = typeof candidate.label === "string" ? candidate.label : "";
    const kind = candidate.kind === "key" ? "key" : "sequence";
    const sequence = Array.isArray(candidate.sequence)
      ? candidate.sequence.filter((entry: unknown): entry is string => typeof entry === "string")
      : [];
    const key = typeof candidate.key === "string" ? candidate.key.trim() : undefined;
    const modifiers = sanitizeModifiers(candidate.modifiers);
    const preset = candidate.preset === true;

    if (!id) continue;
    if (kind === "key" && !key) continue;
    if (kind === "sequence" && sequence.length === 0) continue;

    shortcuts.push({
      id,
      label,
      kind,
      sequence,
      key,
      modifiers,
      preset,
    });
  }

  return shortcuts;
}

function sanitizeQuickCommands(value: unknown): RemoteTerminalQuickCommandData[] {
  if (!Array.isArray(value)) return [];

  const commands: RemoteTerminalQuickCommandData[] = [];

  for (const item of value) {
    if (!item || typeof item !== "object") continue;

    const candidate = item as Partial<RemoteTerminalQuickCommandData>;
    const id = typeof candidate.id === "string" ? candidate.id.trim() : "";
    const label = typeof candidate.label === "string" ? candidate.label.trim() : "";
    const command = typeof candidate.command === "string" ? candidate.command : "";

    if (!id || !label || !command.trim()) continue;

    commands.push({
      id,
      label,
      command,
    });
  }

  return commands;
}

function sanitizePreferences(value: unknown): RemoteTerminalAgentPreferencesData {
  const candidate: Partial<PersistedPreferencesFile> =
    value && typeof value === "object" ? (value as PersistedPreferencesFile) : {};
  const defaultWorkingDirectory =
    typeof candidate.defaultWorkingDirectory === "string" ? candidate.defaultWorkingDirectory.trim() : "";

  return {
    defaultWorkingDirectory: defaultWorkingDirectory || undefined,
    shortcuts: sanitizeShortcuts(candidate.shortcuts),
    quickCommands: sanitizeQuickCommands(candidate.quickCommands),
  };
}

export class AgentPreferencesStore {
  private cache: RemoteTerminalAgentPreferencesData | null = null;

  public constructor(private readonly filePath: string) {}

  public getPreferences(): RemoteTerminalAgentPreferencesData {
    if (!this.cache) {
      this.cache = this.loadFromDisk();
    }

    return clonePreferences(this.cache);
  }

  public setPreferences(preferences: RemoteTerminalAgentPreferencesData): RemoteTerminalAgentPreferencesData {
    const sanitized = sanitizePreferences(preferences);
    mkdirSync(path.dirname(this.filePath), { recursive: true });
    writeFileSync(
      this.filePath,
      JSON.stringify(
        {
          version: PREFERENCES_FILE_VERSION,
          defaultWorkingDirectory: sanitized.defaultWorkingDirectory,
          shortcuts: sanitized.shortcuts,
          quickCommands: sanitized.quickCommands,
        } satisfies PersistedPreferencesFile,
        null,
        2,
      ),
      "utf8",
    );
    this.cache = sanitized;
    return clonePreferences(sanitized);
  }

  private loadFromDisk(): RemoteTerminalAgentPreferencesData {
    try {
      const raw = readFileSync(this.filePath, "utf8");
      return sanitizePreferences(JSON.parse(raw));
    } catch {
      return {
        defaultWorkingDirectory: undefined,
        shortcuts: [],
        quickCommands: [],
      };
    }
  }
}

File: internal/agent/preferences.go
package agent

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/remote-terminal-cloud/agent/internal/protocol"
)

type persistedPreferencesFile struct {
	Version                 int                                       `json:"version"`
	DefaultWorkingDirectory string                                    `json:"defaultWorkingDirectory,omitempty"`
	Shortcuts               []protocol.RemoteTerminalShortcutData     `json:"shortcuts,omitempty"`
	QuickCommands           []protocol.RemoteTerminalQuickCommandData `json:"quickCommands,omitempty"`
}

type PreferencesStore struct {
	filePath string
	mu       sync.Mutex
	cache    *protocol.RemoteTerminalAgentPreferencesData
}

func NewPreferencesStore(filePath string) *PreferencesStore {
	return &PreferencesStore{filePath: filePath}
}

func (s *PreferencesStore) GetPreferences() protocol.RemoteTerminalAgentPreferencesData {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.cache == nil {
		loaded := s.loadFromDisk()
		s.cache = &loaded
	}

	return clonePreferences(*s.cache)
}

func (s *PreferencesStore) SetPreferences(preferences protocol.RemoteTerminalAgentPreferencesData) (protocol.RemoteTerminalAgentPreferencesData, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	sanitized := sanitizePreferences(preferences)
	if err := os.MkdirAll(filepath.Dir(s.filePath), 0o755); err != nil {
		return protocol.RemoteTerminalAgentPreferencesData{}, err
	}

	payload := persistedPreferencesFile{
		Version:                 1,
		DefaultWorkingDirectory: sanitized.DefaultWorkingDirectory,
		Shortcuts:               sanitized.Shortcuts,
		QuickCommands:           sanitized.QuickCommands,
	}
	content, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return protocol.RemoteTerminalAgentPreferencesData{}, err
	}

	if err := os.WriteFile(s.filePath, content, 0o644); err != nil {
		return protocol.RemoteTerminalAgentPreferencesData{}, err
	}

	s.cache = &sanitized
	return clonePreferences(sanitized), nil
}

func (s *PreferencesStore) loadFromDisk() protocol.RemoteTerminalAgentPreferencesData {
	content, err := os.ReadFile(s.filePath)
	if err != nil {
		return protocol.RemoteTerminalAgentPreferencesData{
			Shortcuts:     []protocol.RemoteTerminalShortcutData{},
			QuickCommands: []protocol.RemoteTerminalQuickCommandData{},
		}
	}

	var payload persistedPreferencesFile
	if err := json.Unmarshal(content, &payload); err != nil {
		return protocol.RemoteTerminalAgentPreferencesData{
			Shortcuts:     []protocol.RemoteTerminalShortcutData{},
			QuickCommands: []protocol.RemoteTerminalQuickCommandData{},
		}
	}

	return sanitizePreferences(protocol.RemoteTerminalAgentPreferencesData{
		DefaultWorkingDirectory: payload.DefaultWorkingDirectory,
		Shortcuts:               payload.Shortcuts,
		QuickCommands:           payload.QuickCommands,
	})
}

func clonePreferences(preferences protocol.RemoteTerminalAgentPreferencesData) protocol.RemoteTerminalAgentPreferencesData {
	shortcuts := make([]protocol.RemoteTerminalShortcutData, 0, len(preferences.Shortcuts))
	for _, shortcut := range preferences.Shortcuts {
		sequence := append([]string(nil), shortcut.Sequence...)
		modifiers := append([]protocol.RemoteTerminalShortcutModifier(nil), shortcut.Modifiers...)
		shortcuts = append(shortcuts, protocol.RemoteTerminalShortcutData{
			ID:        shortcut.ID,
			Label:     shortcut.Label,
			Kind:      shortcut.Kind,
			Sequence:  sequence,
			Key:       shortcut.Key,
			Modifiers: modifiers,
			Preset:    shortcut.Preset,
		})
	}

	quickCommands := make([]protocol.RemoteTerminalQuickCommandData, 0, len(preferences.QuickCommands))
	quickCommands = append(quickCommands, preferences.QuickCommands...)

	return protocol.RemoteTerminalAgentPreferencesData{
		DefaultWorkingDirectory: preferences.DefaultWorkingDirectory,
		Shortcuts:               shortcuts,
		QuickCommands:           quickCommands,
	}
}

func sanitizePreferences(preferences protocol.RemoteTerminalAgentPreferencesData) protocol.RemoteTerminalAgentPreferencesData {
	result := protocol.RemoteTerminalAgentPreferencesData{
		DefaultWorkingDirectory: strings.TrimSpace(preferences.DefaultWorkingDirectory),
		Shortcuts:               sanitizeShortcuts(preferences.Shortcuts),
		QuickCommands:           sanitizeQuickCommands(preferences.QuickCommands),
	}
	return result
}

func sanitizeShortcuts(items []protocol.RemoteTerminalShortcutData) []protocol.RemoteTerminalShortcutData {
	result := make([]protocol.RemoteTerminalShortcutData, 0, len(items))
	for _, item := range items {
		id := strings.TrimSpace(item.ID)
		if id == "" {
			continue
		}

		label := item.Label
		kind := item.Kind
		if kind != protocol.ShortcutKindKey {
			kind = protocol.ShortcutKindSequence
		}

		sequence := make([]string, 0, len(item.Sequence))
		for _, entry := range item.Sequence {
			if strings.TrimSpace(entry) != "" {
				sequence = append(sequence, entry)
			}
		}

		key := strings.TrimSpace(item.Key)
		if kind == protocol.ShortcutKindKey && key == "" {
			continue
		}
		if kind == protocol.ShortcutKindSequence && len(sequence) == 0 {
			continue
		}

		modifiers := sanitizeModifiers(item.Modifiers)
		result = append(result, protocol.RemoteTerminalShortcutData{
			ID:        id,
			Label:     label,
			Kind:      kind,
			Sequence:  sequence,
			Key:       key,
			Modifiers: modifiers,
			Preset:    item.Preset,
		})
	}
	return result
}

func sanitizeModifiers(items []protocol.RemoteTerminalShortcutModifier) []protocol.RemoteTerminalShortcutModifier {
	if len(items) == 0 {
		return nil
	}
	result := make([]protocol.RemoteTerminalShortcutModifier, 0, len(items))
	for _, item := range items {
		switch item {
		case protocol.ShortcutModifierCtrl, protocol.ShortcutModifierAlt, protocol.ShortcutModifierShift, protocol.ShortcutModifierMeta:
			result = append(result, item)
		}
	}
	if len(result) == 0 {
		return nil
	}
	return result
}

func sanitizeQuickCommands(items []protocol.RemoteTerminalQuickCommandData) []protocol.RemoteTerminalQuickCommandData {
	result := make([]protocol.RemoteTerminalQuickCommandData, 0, len(items))
	for _, item := range items {
		id := strings.TrimSpace(item.ID)
		label := strings.TrimSpace(item.Label)
		command := item.Command
		if id == "" || label == "" || strings.TrimSpace(command) == "" {
			continue
		}
		result = append(result, protocol.RemoteTerminalQuickCommandData{
			ID:      id,
			Label:   label,
			Command: command,
		})
	}
	return result
}
